### PR TITLE
[mlir][ArmSME] Support 4-way widening outer products

### PR DESCRIPTION
**Pull request by c-rhodes**
This patch introduces support for 4-way widening outer products. This enables
the folding of 4 'arm_sme.outerproduct' operations that are chained via the
accumulator into single widened operations.

Changes:

- Adds the following operations:
  - smopa_wide_4way, smops_wide_4way
  - umopa_wide_4way, umops_wide_4way
  - sumopa_wide_4way, sumops_wide_4way
  - sumopa_wide_4way, sumops_wide_4way
- Implements conversions for the above ops to intrinsics in ArmSMEToLLVM.
- Extends 'arm-sme-outer-product' pass.

For a detailed description of these operations see the
'arm_sme.smopa_wide_4way' description.

### CODE REVIEW

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp [hunk 1]
```diff
@@ -225,37 +271,238 @@ class OuterProductFusion2Way
...
+          op, MATCH_FAILURE_INCONSISTENT_COMBINING_KIND);
+
+    if (!op1->hasOneUse() || !op2->hasOneUse() || !op3->hasOneUse())
+      return rewriter.notifyMatchFailure(
+          op, MATCH_FAILURE_OUTERPRODUCT_NOT_SINGLE_USE);
+
+    if (bool(op1.getLhsMask()) != bool(op2.getLhsMask()) !=
+        bool(op3.getLhsMask()) != bool(op4.getLhsMask()))
+      return rewriter.notifyMatchFailure(op,
+                                         MATCH_FAILURE_INCONSISTENT_MASKING);
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:325]: This does not make sense for four values (and is very confusing to reason about):

`true != false != false != true` => `false`

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp [hunk 2]
```diff
@@ -225,37 +271,238 @@ class OuterProductFusion2Way
...
+    if (
+        // signed, i8i8i32
+        (failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op1, nxnxv4i32, nxv4i8)) ||
+         failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op2, nxnxv4i32, nxv4i8)) ||
+         failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op3, nxnxv4i32, nxv4i8)) ||
+         failed(
+             isCompatible<arith::ExtSIOp>(rewriter, op4, nxnxv4i32, nxv4i8))) &&
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:441]: The expression:
```c++
        (failed(
             isCompatible<Ext>(rewriter, op1, outputType, inputType)) ||
         failed(
             isCompatible<Ext>(rewriter, op2, outputType, inputType)) ||
         failed(
             isCompatible<Ext>(rewriter, op3, outputType, inputType)) ||
         failed(
             isCompatible<Ext>(rewriter, op4, outputType, inputType)))
```
Really should be moved to lambda (or helper), and reused for each case, it's quite hard to follow all the cases here.

Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:441]: A lambda would avoid repeating the arguments again:
```c++
auto failedToMatch = [&](auto extendOp, VectorType resultType, VectorType inputType) {
        using ExtendOpTy = decltype(extendOp);
        return failed(
             isCompatible<ExtendOpTy>(rewriter, op1, resultType, inputType)) ||
         failed(
             isCompatible<ExtendOpTy>(rewriter, op2, resultType, inputType)) ||
         failed(
             isCompatible<ExtendOpTy>(rewriter, op3, resultType, inputType)) ||
         failed(
             isCompatible<ExtendOpTy>(rewriter, op4, resultType, inputType)));
};
```

// using:
```c++
failedToMatch(arith::ExtUIOp{}, nxnxv2i64, nxv2i16)
```

(passing the null `arith::ExtUIOp{}` is a trick to use it as an implicit template parameter, since template proper template lambdas need C++20)

**Review by MacDue — COMMENTED** [review]
Few initial comments:

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp (hunk 1 above)
Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp]: oops, gone now after your suggestion below 👍

[the remark this answers is not in the document]
Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp]: this is a lot cleaner, cheers!

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp (hunk 2 above)
Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp]: this is much better cheers, my c++ is pretty woeful.. 😅

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp [hunk 3]
```diff
@@ -225,37 +271,173 @@ class OuterProductFusion2Way
...
+        rewriter.replaceOpWithNewOp<arm_sme::SuMops4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtSIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UsMops4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        llvm_unreachable("unexpected extend op!");
+    } else {
+      llvm_unreachable("unexpected arm_sme::CombiningKind!");
+    }
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:387]: Use `TypeSwitch`? Similarly to 2-way ;-)

Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:387]: there's two types to switch on because of the signed-by-unsigned / unsigned-by-signed variants and i don't believe it's possible to use `TypeSwitch` for this

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp [hunk 4]
```diff
@@ -225,37 +271,173 @@ class OuterProductFusion2Way
...
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtUIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtSIOp>(lhsExtOp) && isa<arith::ExtUIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::SuMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else if (isa<arith::ExtUIOp>(lhsExtOp) && isa<arith::ExtSIOp>(rhsExtOp))
+        rewriter.replaceOpWithNewOp<arm_sme::UsMopa4WayOp>(
+            op4, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:367]: Why do mixed zero-extended and sign-extened inputs become a unsigned mopa?

Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:367]: > Why do mixed zero-extended and sign-extened inputs become a unsigned mopa?

this is a unsigned-by-signed mopa

Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:367]: Needs some comments :) The difference in names is barely noticeable (and not clear without looking at the definitions)

File: mlir/test/Dialect/ArmSME/outer-product-fusion.mlir [hunk 5]
```diff
@@ -291,6 +894,41 @@ func.func @outerproduct_widening_2way__bad_acc(%a0 : vector<[4]xf16>, %b0 : vect
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__bad_acc
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:897]:
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__missing_acc
```

File: mlir/test/Dialect/ArmSME/outer-product-fusion.mlir [hunk 6]
```diff
@@ -291,6 +894,41 @@ func.func @outerproduct_widening_2way__bad_acc(%a0 : vector<[4]xf16>, %b0 : vect
...
+  %a2_ext = arith.extsi %a2 : vector<[4]xi8> to vector<[4]xi32>
+  %b2_ext = arith.extsi %b2 : vector<[4]xi8> to vector<[4]xi32>
+
+  %a3_ext = arith.extsi %a3 : vector<[4]xi8> to vector<[4]xi32>
+  %b3_ext = arith.extsi %b3 : vector<[4]xi8> to vector<[4]xi32>
+
+  %0 = arm_sme.outerproduct %a0_ext, %b0_ext : vector<[4]xi32>, vector<[4]xi32>
+  %1 = arm_sme.outerproduct %a1_ext, %b1_ext acc(%0) : vector<[4]xi32>, vector<[4]xi32>
+  %2 = arm_sme.outerproduct %a2_ext, %b2_ext acc(%1) : vector<[4]xi32>, vector<[4]xi32>
+  // break chain
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:924]:
```suggestion
  // Missing accumulate - breaks chain of use-def
```

File: mlir/test/Dialect/ArmSME/outer-product-fusion.mlir [hunk 7]
```diff
@@ -314,6 +952,40 @@ func.func @outerproduct_widening_2way__bad_combining_kind(
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__bad_combining_kind
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:955]:
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__inconsistent_combining_kind
```

File: mlir/test/Dialect/ArmSME/outer-product-fusion.mlir [hunk 8]
```diff
@@ -344,6 +1016,41 @@ func.func @outerproduct_widening_2way__cant_erase(
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__cant_erase
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:1019]:
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__multiple_use
```

File: mlir/test/Dialect/ArmSME/outer-product-fusion.mlir [hunk 9]
```diff
@@ -389,6 +1130,41 @@ func.func @outerproduct_widening_2way__bad_masking(
 
 // -----
 
+// CHECK-LABEL: @outerproduct_widening_4way__bad_masking
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:1133]:
```suggestion
// CHECK-LABEL: @outerproduct_widening_4way__inconsistent_masking
```

File: mlir/test/Dialect/ArmSME/outer-product-fusion.mlir [hunk 10]
```diff
@@ -405,6 +1181,37 @@ func.func @outerproduct_widening_2way__bad_defining_op(
   return %1 : vector<[4]x[4]xf32>
 }
 
+// -----
+
+// CHECK-LABEL: @outerproduct_widening_4way__bad_defining_op
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:1186]: Not sure I see what's "wrong" with this example :)

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-i8i8i32.mlir [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+  %a2_ext = arith.extsi %a2 : vector<[4]xi8> to vector<[4]xi32>
+  %b2_ext = arith.extsi %b2 : vector<[4]xi8> to vector<[4]xi32>
+  %a3_ext = arith.extsi %a3 : vector<[4]xi8> to vector<[4]xi32>
+  %b3_ext = arith.extsi %b3 : vector<[4]xi8> to vector<[4]xi32>
+
+  %c1 = arith.constant 1 : index
+  %c2 = arith.constant 2 : index
+  %c3 = arith.constant 3 : index
+  %c4 = arith.constant 4 : index
+
```
Comment (banach-space) [inline mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-i8i8i32.mlir:121]: [nit] Add a comment that with these masks the bottom row of the output matrix is preserved

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp [hunk 12]
```diff
@@ -31,6 +31,55 @@ using namespace mlir;
 using namespace mlir::arm_sme;
 
 namespace {
+
+// Common match failure reasons.
+static constexpr StringLiteral
+    MATCH_FAILURE_NO_ACCUMULATOR("no accumulator operand");
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:37]: Aren't ALL_CAPS reserved for macros? Shouldn't these be camelCase? https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp (hunk 3 above)
Comment (banach-space) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp]: Missed that! Not sure how to improve this then :/

**Review by banach-space — APPROVED** [review]
Some small nits/suggestions, otherwise LGTM, thanks!

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-i8i8i32.mlir (hunk 11 above)
Comment (c-rhodes) [reply mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-i8i8i32.mlir:121]: I've not added this one as that isn't all they do and I think it could be misleading.

File: mlir/test/Dialect/ArmSME/outer-product-fusion.mlir (hunk 10 above)
Comment (c-rhodes) [reply mlir/test/Dialect/ArmSME/outer-product-fusion.mlir:1192]: same as above but for 4-way, i.e. Defining op of outer product must be a supported extension op.

But I've added a comment where it goes wrong to clarify.

**Review by MacDue — APPROVED** [review]
LGTM

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp (hunk 12 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:37]: Just an aside/observation: MLIR is not following the LLVM coding standards exactly, it's a mish-mash of Google style + LLVM style (but follows neither exactly), in both new and old code.

Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:37]: E.g. MLIR is not following the LLVM rules for variable names (nor the Google one for that matter). But it does seem to follow Google constant naming rules.

Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:37]: >Just an aside/observation: MLIR is not following the LLVM coding standards exactly, it's a mish-mash of Google style + LLVM style (but follows neither exactly), in both new and old code.

Yeah I noticed in that doc it says variables start with upper case but that's not true in MLIR. Anyhow, for this particular case, I can't see anywhere else that uses upper case for constexpr like this so I'll change to camel case given it's prevalent.

File: mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp [hunk 13]
```diff
@@ -31,6 +31,65 @@ using namespace mlir;
...
+    matchFailureNoAccumulator("no accumulator operand");
+static constexpr StringLiteral matchFailureExpectedOuterProductDefOp(
+    "defining op of accumulator must be 'arm_sme.outerproduct'");
+static constexpr StringLiteral matchFailureInconsistentCombiningKind(
+    "combining kind (add or sub) of outer products must match");
+static constexpr StringLiteral matchFailureInconsistentMasking(
+    "unsupported masking, either both outerproducts are masked "
+    "or neither");
+static constexpr StringLiteral matchFailureOuterProductNotSingleUse(
+    "outer product(s) not single use and cannot be removed, no benefit to "
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:46]: very nit: It seems more common to prefix constants with `k`, so `kMatch...`

Comment (banach-space) [reply mlir/lib/Dialect/ArmSME/Transforms/OuterProductFusion.cpp:46]: +1
